codec-http2/src/main/java/io/netty/handler/codec/http2/Http2ConnectionHandler.java

```diff
@@ -560,11 +560,15 @@ public ChannelFuture goAway(final ChannelHandlerContext ctx, final int lastStrea
             future.addListener(new GenericFutureListener<ChannelFuture>() {
                 @Override
                 public void operationComplete(ChannelFuture future) throws Exception {
-                    if (!future.isSuccess()) {
+                    if (future.isSuccess()) {
+                        if (errorCode != NO_ERROR.code()) {
+                            ctx.close();
+                        }
+                    } else {
                         String msg = format("Sending GOAWAY failed: lastStreamId '%d', errorCode '%d', " +
                                             "debugData '%s'.", lastStreamId, errorCode, debugData);
                         logger.error(msg, future.cause());
-                        ctx.channel().close();
+                        ctx.close();
```

Comment:
@nmittler @Scottmitch when you call ctx.close() it starts to flow through the ChannelPipeline starting from this ChannelHandler... Channel.close() will start at the tail of the ChannelPipeline and so go through all handlers. Almost 99% of the time you want to use ChannelHandlerContext.close()